expression/builtin_like.go

```diff
@@ -96,15 +96,42 @@ func (c *regexpFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 	bf.tp.Flen = 1
 	var sig builtinFunc
 	if types.IsBinaryStr(args[0].GetType()) || types.IsBinaryStr(args[1].GetType()) {
-		sig = &builtinRegexpBinarySig{bf}
+		sig = &builtinRegexpBinarySig{baseBuiltinFunc: bf, rm: nil}
 	} else {
-		sig = &builtinRegexpSig{bf}
+		sig = &builtinRegexpSig{baseBuiltinFunc: bf, rm: nil}
 	}
 	return sig, nil
 }
 
+type regexpCompiler interface {
+	compile(pat string) (*regexp.Regexp, error)
+}
+
+type regexpMemoizer struct {
```

Comment:
Could we just use the type `*regexp.Regexp` all around instead of creating a new type `regexpMemoizer`?

Reply:
`regexpMemoizer` exists mostly to flag that we've tried to initialize the memoized regex. Otherwise if we just store a pointer to `*regexp.Regexp`, but there is an error when compiling the pattern, then at each invocation of `vecEvalInt()`, we will repeatedly try to compile and fail. 

Alternatively, we can have a `bool` to flag that we have tried the initialization.

Reply:
I prefer to use an error and a pointer to `*regexp.Regexp` to record the result of `compile`.
If the error and the pointer are both nil, then we should do initialization, otherwise one of them should not be nil.

Reply:
Done